### PR TITLE
feat(tabs): prevent tab switch

### LINKED ISSUE #2715: A way to intercept tab change
Hello,

I need a way to intercept a tab change. Here is a situation where such functionality would be useful:

I have two tabs. Each has a form inside. The requirement is that if form in the current tab becomes `dirty`, I need to show a popup message and depending on which option user chooses ((1) "save changes", (2) "discard changes", (3) "cancel") I should either (1) save and switch tab, (2) don't save and switch tab, or (3) don't save and stay on current tab.

Please let me know if that functionality already exists. If it does not, what approach do you think would be most suitable?

### PR DESCRIPTION
**Pull request by larvanitis**
Closes #2715.

### CODE REVIEW

**Comment by larvanitis** [comment]
Fixed missing $event in scope.select

**Comment by rvanbaalen** [comment]
Seems like some good work to me. Can you also add some tests for this new feature?

File: template/tabs/tab.html [hunk 1]
```diff
@@ -1,3 +1,3 @@
 <li ng-class="{active: active, disabled: disabled}">
-  <a href ng-click="select()" tab-heading-transclude>{{heading}}</a>
+  <a href ng-click="select($event)" tab-heading-transclude>{{heading}}</a>
```
Comment (wesleycho) [inline template/tabs/tab.html]: With this change, documentation should be added for the `select` option as well.

Comment (larvanitis) [inline template/tabs/tab.html]: This `select` is only used internally, and the event is passed to `beforeSelect` expression which is documented.

Do you mean document the `scope.select` function in [src/tabs/tabs.js](https://github.com/larvanitis/bootstrap/blob/prevent-tab-switch/src/tabs/tabs.js#L229) using JSDoc?

Comment (wesleycho) [inline template/tabs/tab.html]: Ah, ignore my comment, I was mistaken.

**Comment by wesleycho** [comment]
This just needs tests.

I still want to think whether this is the best approach, but otherwise this is generally ok.

**Comment by wesleycho** [comment]
Any luck with tests @larvanitis?

**Comment by larvanitis** [comment]
Sorry mate. I really don't have any free time yet. Could you have a take at tests?

**Comment by wesleycho** [comment]
Shouldn't be too hard to write I think - in the `basics` describe section, one just needs something like
```javascript
it('should cancel selection of the tab', function() {
  scope.beforeSelect = function($event) {
    $event.preventDefault();
  });
  spyOn(scope, 'beforeSelect').and.callThrough();
  elm = $compile([
    '<tabset class="hello" data-pizza="pepperoni">',
    '  <tab heading="First Tab {{first}}" active="actives.one" select="selectFirst()" deselect="deselectFirst()">',
    '    first content is {{first}}',
    '  </tab>',
    '  <tab active="actives.two" select="selectSecond()" before-select="beforeSelect($event)" deselect="deselectSecond()">',
    '    <tab-heading><b>Second</b> Tab {{second}}</tab-heading>',
    '    second content is {{second}}',
    '  </tab>',
    '</tabset>'
  ].join('\n'))(scope);
  expect(scope.actives.one).toBe(true);
  expect(scope.actives.two).toBeFalsy();

  titles().eq(1).find('a').click();
  expect(scope.beforeEach).toHaveBeenCalled();
  expect(scope.actives.one).toBe(true);
  expect(scope.actives.two).toBeFalsy();
  expect(scope.deselectFirst).not.toHaveBeenCalled();
  expect(scope.selectSecond).not.toHaveBeenCalled();
});
```

I think this would be a sufficient test.

**Comment by wesleycho** [comment]
The PR also needs to be rebased due to code style fixes in `master`.

**Comment by wesleycho** [comment]
Closing this as outdated - feel free to open a new PR reimplementing this based on `master` if you feel it is a sensible feature

**Pull request closed without merging**